add UpdateComment mutation

ALau2088 opened the pull request:
## Pullrequest
Functionality for users to edit their comments.

### Issues
<!-- Which Issues does this fix, which are related?
- fixes #552
- relates #XXX
-->
- [X] None

### Checklist
<!-- Anything important to be thought of when deploying?
- [ ] Env-Variables adjustment needed
- [ ] Breaking/critical change
-->
- [X] None

### How2Test
<!-- Give a detailed description how to test your PR and confirm it is working as expected. -->
<!-- Maintainers will check the Tests
- [ ] Test1
- [ ] Test2
-->
- [X] None

### Todo
<!-- In case some parts are still missing, list them here. -->
- [X] None

ALau2088 commented:
@mattwr18 Here is what I have so far. Still work in progress. For some reason I get an empty records array from neo4j when trying to match the comment. Is this a frontend and backend ticket or just ticket for the backend?

ALau2088 commented:
@mattwr18 I am interested in implementing both frontend and backend but I just wanted to confirm so I am not stepping on anyone's toes.

ALau2088 commented:
#552

ALau2088 commented:
@mattwr18 Got the backend working now. Let me know if I can work on the front end. Thanks.

mattwr18 commented on backend/src/resolvers/comments.js:
```diff
@@ -36,20 +40,84 @@ export default {
+          userId: context.user.id,
+          postId,
+          commentId: comment.id
+        }
```
this is great @roschaefer, thanks for this :smile_cat:

mattwr18 commented on backend/src/resolvers/comments.js:
```diff
@@ -36,20 +40,84 @@ export default {
+        throw new UserInputError(
+          `Comment must be at least ${COMMENT_MIN_LENGTH} character long!`
+        )
+      }
```
yes @ALau2088, as @roschaefer said, we are trying to add all our backend validations to the validations middleware, which @roschaefer has newly added. 

We need to create a ticket to refactor all backend validations to this new middleware. Maybe something you could help out with @ALau2088?

mattwr18 commented on backend/src/resolvers/comments.js:
```diff
@@ -36,20 +40,84 @@ export default {
+          content
+        }
+      )
+      session.close()
```
I'd be interested to know more about what you are talking about @roschaefer, I'm not sure I understood what you are trying to say... :thinking: 

It would be great to see some backend tests though @ALau2088, but so far it's looking good... great job!!

roschaefer replied:
@ALau2088 first checks if the comment exists, then sending an error message in case, then updating the comment, then `MATCH`ing the comment again to update the `content` - I find it cumbersome and I wanted to suggest to check if there is an easier way?

Why not simply:

```
      params.content = params.content.replace(/<(?:.|\n)*?>/gm, '').trim()
```
at line 69?

I guess you can remove everything after
```
await neo4jgraphql(
        object,
        params,
        context,
        resolveInfo,
        false
      )
```

mattwr18 replied:
great! that makes sense @roschaefer

mattwr18 commented:
> @ALau2088 you are certainly not stepping into anybody's toes. We practice collective code ownership and we appreciate if people from the community help out on tasks. So:
> 
> ![You are awesome](https://camo.githubusercontent.com/a1cef542288d5ee97b7a0df87b2394b65570b7e4/68747470733a2f2f6d656469612e67697068792e636f6d2f6d656469612f7854394450724331595758424f31636f79512f67697068792e676966)
> 
> Did you find your way to our community chat already? https://discordapp.com/invite/DFSjPaX

I agree with @roschaefer! I think it's really nice to have developers who want to deliver features, including the frontend, backend, and testing. Of course, not everyone is comfortable doing this, but please feel free to continue on this `PR`

@roschaefer, @ALau2088 is both in the Discord chat and in the AV slack, and we've had a few pairing sessions already. :smile_cat:

ALau2088 commented:
Thanks for the feedback @mattwr18 and @roschaefer! 
I have updated the pull request. Here is a brief summary list of the changes:
-Refactored comment validations
-Refactored UpdateComment mutation per @roschaefer comments.
-Added edit comment UI.
-I still have the testing to do. Hoping to get to it this weekend. To be continued...

mattwr18 commented:
That's great @ALau2088, I was taking a look yesterday! Seems like great progress!

I'll pull down the branch and play around with it :smile_cat: 

Currently it's failing the build cause of linting issues, if you get in the habit of running `yarn lint --fix` from the `webapp` and `backend` directories, it won't fail the build because of them ;)

Also, there has been merged in some config for vscode users to have auto formatting on save, let us know if it's working for you

ALau2088 commented on backend/src/middleware/validation/index.js:
```diff
@@ -22,10 +22,29 @@ const validateUrl = async (resolve, root, args, context, info) => {
+    throw new UserInputError(NO_POST_ERR_MESSAGE)
+  }
+
+  return await resolve(root, args, context, info)
```
@mattwr18 I am getting an ES-Linting error on line 39. error  Redundant use of `await` on a return value  no-return-await. Did you encounter this before? Any suggestions for a workaround?

mattwr18 replied:
maybe you need to save the returned value from the `await resolve(root, args, context, info)` to a variable and return that?

ulfgebhardt replied:
`return` automatically `awaits` therefore its redundant.
The Solution would be `to return resolve(...)` I guess

ALau2088 replied:
@mattwr18 @ulfgebhardt Thanks for the input guys!

roschaefer replied:
@ALau2088 it seems that `return await resolve` is in fact different from `return resolve`. I guess, 99.9% of all cases it should make no difference and if it does, it's probably a code smell.

Nevertheless, I experiences issues when I removed the `await` [here](https://github.com/Human-Connection/Human-Connection/blob/master/backend/src/middleware/validation/index.js#L7). So if you *want* disable the linter, you can do so by adding a comment (see the linked LOC above :point_up: ).

ALau2088 commented:
@mattwr18 Is there a certain test script I should use for testing individual spec files?

mattwr18 commented:
@ALau2088, in the `/webapp` directory, you can run `yarn test <relative_path_to_spec>`
from the `/backend` directory, you can run `yarn test:jest <relative_path_to_spec>`

ALau2088 commented:
@mattwr18 Still unable to get the test to run. Here is a snapshot of sample test I tried to run.
![snapshot](https://user-images.githubusercontent.com/37641224/58669552-78fbae00-82f1-11e9-8a43-65795716971b.png)

mattwr18 commented:
hey @ALau2088... yeah sorry about that, our scripts are not really friendly to `Windows` users... you might want to try using `Docker`?? then you will run the tests inside a container, which uses Linux...

Alternatively, I know that @ulfgebhardt has Windows and is able to run the tests... could you share your adaptation of the scripts @ulfgebhardt??

roschaefer commented:
@ALau2088 it might be that on Windows, the paths are different. Did you try prepending the path with a `./`? On UNIX that's short for "current folder". With a little luck, that works on Windows too.

mattwr18 commented:
I think the solution that I saw @ulfgebhardt use involved removing these `2> /dev/null` from `test:before:server` and `test:before:seeder`, but there was some other trick as well, that, to be honest, I didn't pay close enough attention to...

ulfgebhardt commented:
@ALau2088 Yes you have to remove `2> /dev/null` and you are not allowed to commit that change since else the test runner spams alot of errors (refactor pls @roschaefer ) -.-
Also it is not possible to run single tests under windows(or at least not out of the box) you will always have to run the whole test package

ALau2088 commented:
![2](https://user-images.githubusercontent.com/37641224/58999921-16a52080-87bd-11e9-9d6b-6d2ede3cd607.png)
Hey @mattwr18 . I was thinking of refactoring CreateComment and UpdateComment into a single describe maybe call it "Comment Functionality" to avoid duplicating variables in the describe('authenticated') but wanted to see what you think first. Thanks.

ALau2088 commented on webapp/components/comments/EditCommentForm/index.vue:
```diff
@@ -0,0 +1,121 @@
+        .then(res => {
+          this.loading = false
+          this.$root.$emit('refetchPostComments')
+          this.$toast.success(this.$t('post.comment.submitted'))
```
@mattwr18 . Do I pull down the latest by running _git pull upstream 552-update_comment_ ?

ALau2088 replied:
@mattwr18 . The post.comment.updated has been added.

mattwr18 commented:
that sounds like a really good idea @ALau2088... most importantly though, did you read my comment about pushing directly to the human-connection repo? it's necessary for the build to pass!

ALau2088 commented:
> that sounds like a really good idea @ALau2088... most importantly though, did you read my comment about pushing directly to the human-connection repo? it's necessary for the build to pass!

@mattwr18 . Is that done by running the command _'git push upstream 552-update_comment'_? Before I do that do I need to run _'git pull upstream 552-update_comment'_ to pull down your edits?

mattwr18 commented:
Almost @ALau2088, first you would run `git pull origin 552-update_comment` because I pushed up to your forked branch. Then, `git pull upstream master` to bring in the latest changes from `master` then `git push upstream 552-update_comment`, and open a `PR`.

Then you can either close this, or I can. :smile_cat:

ALau2088 commented:
![3](https://user-images.githubusercontent.com/37641224/59071613-47945c80-8874-11e9-9269-ffc26b104253.png)
@mattwr18 Any thoughts on why updateComment is null?

mattwr18 commented:
I'd have to look at the code... have you pushed up the latest?

mattwr18 commented on backend/src/resolvers/comments.spec.js:
```diff
@@ -78,7 +78,37 @@ describe('CreateComment', () => {
+      let updateCommentVariables = {
+        postId: 'p1',
+        content: 'Comment is updated',
+        id: 'c8'
```
do you have a comment with the id `c8`... I can't see where you are creating that...

roschaefer commented on webapp/locales/en.json:
```diff
@@ -145,7 +145,8 @@
       "submit": "Comment",
-      "submitted": "Comment Submitted"
+      "submitted": "Comment Submitted",
+      "updated": "Comment Updated"
```
@ALau2088 the German translation for `Commented Updated` would be `Kommentar Aktualisiert`. Although I have to admit this sounds a little strange to me. Probably I would use a more generic wording `Changes Saved` <-> `Änderungen gespeichert`.
